packages/coreutils/src/json.ts

```diff
@@ -212,29 +262,44 @@ namespace JSONExt {
   /**
    * Compare two JSON objects for deep equality.
    */
-  function deepObjectEqual(first: ReadonlyJSONObject, second: ReadonlyJSONObject): boolean {
+  function deepObjectEqual(first: ReadonlyPartialJSONObject, second: ReadonlyPartialJSONObject): boolean {
     // Check referential equality first.
     if (first === second) {
       return true;
     }
 
     // Check for the first object's keys in the second object.
     for (let key in first) {
-      if (!(key in second)) {
+      if (first[key] !== undefined && !(key in second)) {
```

Comment:
I don't think this is needed. Mismatched `undefined` values will be caught with the equality check below.

Reply:
Removing this fails the test `should handle optional keys on partials`. The failing case is `JSONExt.deepEqual( { bar: undefined }, {})`. Here, the key `bar` will be present in `first`, but not in `second`, and they will compare as unequal.